### ISSUE TITLE
For faction bases make it so you can send more people on construction mission to drastically cut time

### ISSUE THREAD

**Issue by hostergaard** [CleverRaven/Cataclysm-DDA#53656]
### Is your feature request related to a problem? Please describe.

The time requirement for building things are so immense compared to the general game speed that often, before I manage to actually get to a point where I can enjoy the benefits of the base I am long done and bored with the save. It litterally take years to expand the base. Which is realistic if you are one person doing it but just having two people would more than halve the time it takes cause when you are alone, a lot of tasks are exponentially more difficult, but with one to support you can do tasks far more efficiently. I say that as a son of carpenter that have build and pulled down multiple houses. Alone just lifting a beam up takes a while as you have to go back and forth and use a lot of tricks, but with a partner you can skip a lot of extra work. 

### Solution you would like.

Make you able to add more people to every construction people. Adding more people should decrease it at least linearly if not more up until a certain point. I.e. 10 days should be 5 days for 2 people, 3.33 for 3 people, 2.5 for 4 people and so on. 

### Describe alternatives you have considered.

- Powered/higher quality tools decreasing time compared to unpowered tools

- A powered forklift arm drastically reducing build time to simulate the use of machinery in building 

These solutions might also be wise to implement along with the number of people reduction. Its reasonable that a well functioning base with lots of people, powered tools and large machinery would able to construct things extremely quickly.

**EDIT:** Per Photoloss insightfull comment I made expanded on what could be a better solution:

I like what you said about reworking it into blueprints, then it would give the player the opportunity to affect the speed with all the methods available to them in game instead of some abstraction. In fact, going alone those lines maybe it should work closer to how zones work, so the faction camp menu allow you can set up any arbitrary zone and simply checks if the required items, walls, furniture and so is within those zones allowing the player to access the recipes and productions of basecamp.

And then detach blueprint design into a separate selection, where players can select blueprints to apply but are not necessary or linked to the function of the base per se and more about how you could design building. These blueprints ideally would be importable and exportable so users can share base designs and work out particular design for bases in various in game buildings.

So to sum up, for base functions it should basically just work like zones and just check whatever required is within that zone or zones, like how base storage already works. And then actual construction should work trough a importable and exportable blueprints. That way the players can design the bases however they want, or use pre designed designs but the functioning of the base stays more or less the same (making a labyrinth probable would impact the game some). 

### Additional context

https://www.youtube.com/watch?v=AsTB0HnM6WM

A bunch of Amish constructing a entire barn in a day using their hands and primitive tools.

**Comment by Photoloss** [comment]
While not as visually impressive NPCs can already do this via "construction blueprint" zones. Making that feature easier to apply in bulk would benefit constructions not tied to faction camps as well. And superior tools should definitely improve the player's personal construction projects anyway. The faction camp "mission" could be reworked to simply place a pre-made blueprint set, possibly with options for moving it in a placement mode similar to the debug tile editor. This should also allow the player to chip in to speed things up. Ideally faction camps would work based on factual availability, checking the overmap squares for enclosed rooms and furniture rather than relying on abstract "mission" progression which to my knowledge is not lost if the physical structure burns down afterwards either.

**Comment by hostergaard** [comment]
Yeah, its specifically the base camp "missions" that are the issue, as it takes far too long for most people to enjoy I believe and with no opportunity to  affect how quickly it goes.

I like what you said about reworking it into blueprints, then it would give the player the opportunity to affect the speed with all the methods available to them in game instead of some abstraction. In fact, going alone those lines maybe it should work closer to how zones work, so the faction camp menu allow you can set up any arbitrary zone and simply checks if the required items, walls, furniture and so is within those zones allowing the player to access the recipes and productions of basecamp. 

And then detach blueprint design into a separate selection, where players can select blueprints to apply but are not necessary or linked to the function of the base per se and more about how you could design building. These blueprints ideally would be importable and exportable so users can share base designs and work out particular design for bases in various in game buildings. 


So to sum up, for base functions it should basically just work like zones and just check whatever required is within that zone or zones, like how base storage already works. And then actual construction should work trough a importable and exportable blueprints. That way the players can design the bases however they want, or use pre designed designs but the functioning of the base stays more or less the same (making a labyrinth probable would impact the game some).

**Comment by PatrikLundell** [comment]
The longest base camp construction sequence builds out of concrete, and it takes about two seasons for a single companion to complete the job for the field base camp itself. However, the current logic allows companions to work on different sections in parallel, and you can start expansions to allow for even more parallelism (my construction rate was limited by the rate at which I could get construction material, not the work force, which mostly consisted of the starting companion, the refugee camp beggars, and a rescued cyborg). True, there are some construction sequences that are locked to piggy back onto each other, but many are not (in particular for the version 2 variants, where I've tried to decouple things, so essentially only shared walls serve as prerequisites [the "furniture" picture is a bit more complicated, but also done a lot faster]).
I've had my all metal (only version at the time, version 1 now) base camp with a full set of expansions constructed before winter with a standard spring start in 0.E2 stable. Thus, it taking years is an exaggeration.

However, I fully agree that having the game scan what functionality is available based on what's on the tiles would be superior to having to have blueprints tell the game what it should provide. However, I believe there's still a use for camp blueprints for those who want to build base camps, but not play architect, but, ideally, they should only be an available resource, not a requirement.

I also think that converting blueprints into a set of orders to construct each tile as a separate task to be picked up by companions would be a good thing. Note, though that that would require the game to have a logic that blocks the construction of elements that require the support of other elements (so you can't construct the floor/wall in the middle of a room before you've got support for the ceiling). Also note that some existing blueprints are actually impossible to construct by hand (the livestock expansion contains a real world logical pillar supported structure that the current support logic doesn't allow), and thus would be invalidated. Apart from the work to split the blueprint jobs into pieces, there would also have to be game support for companions actually working on the pieces, including when the reality bubble moves away from them. Vehicle dismantling might provide some clues as how to do that (minus the crash inducing usage of mutable indices).

Having the "crafting" of anything depend on the actual tools used when it comes to the time and effort used is something that's currently not supported (using a powered chain saw to cut down a tree takes the same time and stamina as using a hand saw, plus it uses up power, for instance). There's definitely room for improvement in that department.

Construction of blueprint tiles using multiple companions should probably use the same logic as other crafting, although if blueprints are broken down into individual tiles there's probably a limited need for piling workers onto the same tile (even though some tasks are easier and faster to perform in groups even when the time used by all the individuals is summed up in the real world).

**Comment by SunshineDistillery** [comment]
The crumminess of basecamps is off topic. They exist in the game now. The original suggestion is a good one and a quicker improvement than at total rework.

**Comment by hostergaard** [comment]
>... Thus, it taking years is an exaggeration.

That is true probably, its more that I several times have build them from scratch and found it to be exceedingly time consuming. But more so in real life hours. It takes literall real life days even if I use the console commands to spawn the needed items, npc's and simply stand on my bed, food and water spawned and wait and sleep as much as I can. While I love a challenge I only have so much life time to spend on it and it gets boring. 
> 
> However, I fully agree that having the game scan what functionality is available based on what's on the tiles would be superior to having to have blueprints tell the game what it should provide. However, I believe there's still a use for camp blueprints for those who want to build base camps, but not play architect, but, ideally, they should only be an available resource, not a requirement.

Exactly, we should decouple the functionality and the design of bases, even tough one could argue they support each other to a degree. ideally we should also be able to export and import these blueprints so people who don't like playing architects can enjoy what people who do design. I love seing those pre designed bases. 

> 
> I also think that converting blueprints into a set of orders to construct each tile as a separate task to be picked up by companions would be a good thing. Note, though that that would require the game to have a logic that blocks the construction of elements that require the support of other elements (so you can't construct the floor/wall in the middle of a room before you've got support for the ceiling)...

Well, here the whole, building in stages can come into play and more importunately a real world thing currently missing in game; scaffolding. 

> 
> Having the "crafting" of anything...

I really should get into figuring out the code and see if I can do some coding there.

> 
> Construction of blueprint tiles using multiple companions should probably use the same logic as...

Might be an idea to allow NPC's to craft from a distance to a degree. A few tiles away would probably help immensely.

**Comment by Photoloss** [comment]
> That is true probably, its more that I several times have build them from scratch and found it to be exceedingly time consuming. But more so in real life hours. It takes literall real life days even if I use the console commands to spawn the needed items, npc's and simply stand on my bed, food and water spawned and wait and sleep as much as I can. While I love a challenge I only have so much life time to spend on it and it gets boring.

While by no means a proper solution one of the debug mutations (probably Crafting Virtuoso) lets you recall followers from those missions instantly after starting as a successful completion. You'll need to make your own honour rules for cooperation, skill impact etc. and do the math yourself but it's better than giving up in frustration especially while testing the actual mapgen of the camps.

**Comment by PatrikLundell** [comment]
I can use the debug facilities to construct a complete base camp or expansion from scratch in one material in about half an hour in real time to test it out (I've done it far too many times to test them out). Rather than sit around to wait for construction to finish you use the debug facility to advance time, which means: spawn materials, order the task to be done, pass X days to allow for enough time to pass, invoke the bulletin board again to recall the companion, access the bulletin board to order the construction of the new segment (possibly after having spawned additional construction materials).

However, in a real play through construction is a background activity I send companions off on while my character engages in exploration, fighting, materials collection (for further construction). As mentioned before, I send companions off to work on things in parallel when I have enough materials. I don't find construction times to be unrealistic (although I don't have any useful real world construction experience). I spend a lot of real time on activities that takes a lot of play time (crafting can take a lot of time, causing the calendar days to pass by fairly quickly, though). I consider base construction to be a process akin to growing crops, i.e. something that's kept going in the background and eventually will complete, rather than something I sit around waiting for.

Scaffolding adds more tool materials to keep track of and would be of more value for construction on additional Z levels. I'm not sure what benefit it would provide for ground level construction. I think there's a need to describe what it would add to the game by introducing it (and what functionality would need to be added, as just adding the need for additional non consumed materials probably isn't that attractive).

**Comment by hostergaard** [comment]
Its not so much a question of it being realistic as it is a question of being functional, I have tried constructing bases several times in game, and with the slow speed days goes by even crafting playing the my normal game, but it still takes too long, I tried the waiting thing as test to see how long it took with no delays between. I litterally sat several whole days running the game trying to progress trough it as fast as humanly possible, with there being no way to speed up construction its effectively useless to the vast majority of players. And its not because my computer is slow or that I have that large a save, it simply takes too much time even left in background. By the time its up and running most games are long over or dead from save bloat. 

The time is realistic enough for one person, but the problem is that the game does no allow two or more people to do it. The benefit of doing it with more than one is not just linear, its more. Two people building one house each will take much longer than two people building two houses together in a row in terms of total work hours. Right now its too much of a bottleneck without any way to escape it. What I want is a way for the player to escape the bottleneck.

And that would be quite simple, send more than one person on a building mission. 

Scaffolding would be a way to prevent the issue of floating construction, you don't really need new materials per se, planks or pipes would do with nails or threads. Something that is easy and quick to set and pull down with no material loss.

**Comment by PatrikLundell** [comment]
If you have the problem that you run out of game after half a year or so faster base building is not going to help with that: you're just going to have some additional buildings with limited use when you run out of game. I see it as a game time extender (and a way to make use of the otherwise rather useless companions), rather than a game chore, by providing you with something to do when you've got the basics covered. If you don't want to spend time on construction, however, skip the boring part (which probably includes hauling back construction materials) and go for the parts you like.

I agree in most situations building construction should get a big boost from the cooperation of several people, but the gain varies depending on the task, so either you end up with a simplistic model where adding more people splits the time by dividing the total time by something only slightly less than the number of people (indicating waiting for tasks to be finished, time lost due to actually having to coordinate activities, and so on) or a rather complicated one where each task would need some model of how adding people would affect it, including how adding more people beyond the optimum number affects it, plus logic to determine how the people allocated to a blueprint would be distributed over time to be used to the best effect. Note, though, that allowing multiple people to help out with tasks might require the time taken for some tasks to need to be increased for the single character case if a more complex model is used, which probably wouldn't be well received by people who want to do without companions.
Regardless of the underlying model, allowing the UI to send off more people on the same task would probably take care of that part.
I still don't think it's the optimal use of developer efforts, but I certainly won't complain if you were able to convince someone to implement it (unless the underlying model is completely bonkers, of course). If blueprint construction gets changed from abstract construction to actual construction tile by tile (which is the direction I would like to see things head in) blueprint level construction management will probably have to be replaced with orders to set up tasks to construct the "terrain" for the individual tiles, and all the blueprint management level stuff removed.

Scaffolding: It's probably not obvious, but I used the term "tool materials" to indicate the use of materials as non consumed "tools". There are current recipes that make use of this, e.g. a log for earthern walls (if I remember correctly) and a wooden frame for some (other?) wall.
However, note that all recipes are on a per tile level, so there is no support for anything on a larger scale than that, e.g. the scaffolding for a blueprint would be the scaffolding for the most "expensive" tile for each tool material. It is POSSIBLE to override the (rather convenient) logic that calculates requirements in terms of materials, skills, and time from each tile in the blueprint, but it's an all-or-nothing thing, i.e. in order to add scaffolding tool materials you'd have to manually calculate everything else and specify it (including updating it when things change, as has happened with the recent introduction of welding rods).
